pkg/roll/run_update.go

```diff
@@ -57,65 +58,57 @@ func NewUpdate(
-	// early. Run can be invoked again to resume after termination, if it did
-	// not lose exclusivity of the RCs.
-	Run(quit <-chan struct{}) error
+	// should claim exclusive ownership of both affected RCs, and release that
+	// exclusivity upon completion. Run is long-lived and non-blocking; close
+	// the quit channel to terminate it early. If an Update is interrupted, Run
+	// should leave the RCs in a state such that it can later be called again to
+	// resume. The error channel returned from Run() must be drained.
+	Run(quit <-chan struct{}) <-chan error
 }
 
-func (u update) Prepare() error {
-	u.logger.NoFields().Debugln("Locking")
-	if err := u.lockRCs(); err != nil {
-		return err
-	}
-
-	u.logger.NoFields().Debugln("Enabling/disabling")
-	if err := u.enable(); err != nil {
-		return err
-	}
+func (u update) Run(quit <-chan struct{}) <-chan error {
+	ret := make(chan error)
+	go u.run(quit, ret)
+	return ret
+}
 
-	newFields, err := u.rcs.Get(u.NewRC)
-	if err != nil {
-		return err
-	}
-	checks, err := u.hcheck.Service(newFields.Manifest.ID())
-	if err != nil {
-		return err
+func (u update) run(quit <-chan struct{}, errs chan<- error) {
+	u.logger.NoFields().Debugln("Locking")
+	// TODO: implement API for blocking locks and use that instead of spamming
+	for err := u.lockRCs(); err != nil; err = u.lockRCs() {
+		errs <- err
+		// in case of fast failure (eg network problem), this sleep will avoid
+		// massive error spinning
+		time.Sleep(1 * time.Second)
 	}
 
-	newNodes, err := u.countHealthy(u.NewRC, checks)
-	if err != nil {
-		return err
-	}
-	oldNodes, err := u.countHealthy(u.OldRC, checks)
-	if err != nil {
-		return err
-	}
+	// release both locks on termination
+	defer func() {
+		// failure to unlock probably means:
+		// - session was already invalidated, lock was already released when
+		//   that happened
+		// - network error, session will probably time out soon
+		if err := u.lock.Unlock(u.lockPath(u.OldRC)); err != nil {
+			errs <- err
+		}
+		if err := u.lock.Unlock(u.lockPath(u.NewRC)); err != nil {
+			errs <- err
+		}
+		close(errs)
+	}()
 
-	if u.DesiredReplicas < u.MinimumReplicas {
-		return fmt.Errorf("desired replicas (%d) less than minimum replicas (%d), update is invalid", u.DesiredReplicas, u.MinimumReplicas)
-	}
-	if newNodes.healthy+oldNodes.healthy <= u.MinimumReplicas {
-		return fmt.Errorf("minimum (%d) is not satisfied by existing healthy nodes (%d old, %d new), update would block immediately", u.MinimumReplicas, oldNodes.healthy, newNodes.healthy)
+	u.logger.NoFields().Debugln("Enabling")
+	for err := u.enable(); err != nil; err = u.enable() {
+		errs <- err
+		time.Sleep(1 * time.Second)
 	}
-	return nil
-}
 
-func (u update) Run(quit <-chan struct{}) error {
+	u.logger.NoFields().Debugln("Launching health watch")
 	newFields, err := u.rcs.Get(u.NewRC)
-	if err != nil {
-		return err
+	for ; err != nil; newFields, err = u.rcs.Get(u.NewRC) {
+		errs <- err
+		time.Sleep(1 * time.Second)
```

Comment:
ugh I knew I forgot something, I'll fix this and find a way to make the retry loops less repetitive